Add Service Now

pzubaty opened the pull request:
It's done. You need ServiceNow REST API user with password to use it. Also add_attachment is missing.
@dmranck @rpavelka @rbosak @amisstea

radomirbosak commented:
Should we review this, @pzubaty ? Or is it just FYI ?

pzubaty commented:
It's mainly FYI, but you can always comment.

rpavelka commented on README.rst:
```diff
@@ -40,43 +40,49 @@ Installation
+variable named TICKETUTIL_DEBUG to 'True'. If this environment variable
+is set to anything else or does not exist, debug logging will be
+disabled.
+
```
^ This should be merged with "Note on the logging" below. I would prefer this to be removed and you to use original. IMHO this even is the case as I can not find any mention of `TICKETUTIL_DEBUG`:

```
(branch: add_verify_project) $ grep TICKETUTIL_ ticketutil/*.py
ticketutil/ticket.py:LOG_LEVEL = os.environ.get('TICKETUTIL_LOG_LEVEL', 'INFO')
```

dmranck replied:
Yes, this can be deleted, as it's been overwritten by the "Note on logging"

rpavelka commented on docs/servicenow.md:
```diff
@@ -0,0 +1,173 @@
+
+This document contains information on the methods available when working
+with a ServiceNow object. A list of the ServiceNow fields that have been
+tested when creating and editing tickets is included. Because each
```
I would start a new paragraph from "Because each ...". I would also add "Note: " before "Because...".

IMHO it would be easier to  read and feel little bit easier also :-) However this is just my feeling and feel free to ignore this.

rpavelka commented on docs/servicenow.md:
```diff
@@ -0,0 +1,173 @@
+category = 'Category'
+item = 'ServiceNow'
+contact_type = 'Email'
+opened_for = 'dept'
```
`dept` value is not explaining purpose of the field. I think it is name of user opening this, right? Then something like `opened_for = 'requesting user'` or something like this would make sense...

rpavelka replied:
Or did I get it wrong and these are real fields of the ticket created by the code above? Then just change the line "Tested create() ticket fields:" to "Ticket created in one particular ServiceNow instance by the line above has been created with following fields." But this does not make sense to me because of some of the values of the fields below...

rpavelka replied:
Ah I see, I get it wrong for the second time too. These are values you have tested to add as an keyword arguments to create command, right? Maybe "Tested create() ticket fields:" to "Following keyword arguments were tested and accepted by our particular ServiceNow instance during ticket creation:"

rpavelka commented on docs/servicenow.md:
```diff
@@ -0,0 +1,173 @@
+
+### get_ticket_content(self, ticket_id) <a name="content"></a>
+
+Retrieves ticket content inside a dictionary. Optional parameter ticket_id
```
`s/inside/as/` or maybe `s/inside/in/`. But I was always bad at languages and maybe I am totally wrong here...

pzubaty replied:
No, you are right, "content as a dictionary" makes more sense.

rpavelka commented on docs/servicenow.md:
```diff
@@ -0,0 +1,173 @@
+
+### devops_one_url(server, table, sys_id) <a name="devops_one"></a>
+
+This function is not part of the ServiceNowTicket, but it's related. It creates
```
This function is not part of the `ServiceNowTicket` class, but it is provide in the same module. But I would prefer this to be part of the `ServiceNowTicket` class. Then you can use `servicenow_url` and `sys_id`. `table_name` and `project` is the same thing? Then I would recommend to use the same name too.

rpavelka commented on ticketutil/servicenow.py:
```diff
@@ -0,0 +1,410 @@
+    Creates DevOps One URL of the existing ticket
+    """
+    return '{server}/pnt/?id=ticket&sys_id={sys_id}&table={table}'.format(
+            server=server, sys_id=sys_id, table=table)
```
I would repeat myself here: I would prefer this to be part of the ServiceNowTicket class. Then you can use `servicenow_url` and `sys_id` attributes. `table_name` and `project` are the same thing? Then I would recommend to use the same name too.

rpavelka replied:
Hmm, I have got another thought on this. This is one particular instance used at one particular company. May you generalize to any such custom instance or maybe remove this? @dmranck any thoughts on this?

dmranck replied:
@rpavelka I agree that this should be removed. I think someone using ticketutil would have this function in their code as a helper function, but it doesn't need to be included in the ticketutil code itself. Because ticketutil should contain general functions that should be able to interact with any instance of the supported tools. ticketutil should not contain specific functions that only interact with our internal instances of the tools.

pzubaty replied:
The problem is that ticketutil uses self.project, but for Service Now itself table makes more sense, but I have renamed it to `self.project`.

rpavelka commented:
@pzubaty how to run unittests provided? In which directory and by what command? I have tried few variations on `python3 -m unittest test_servicenow.py` in test directory but without any success.

rpavelka commented on ticketutil/servicenow.py:
```diff
@@ -0,0 +1,410 @@
+        except requests.RequestException as e:
+            logging.error("Error while getting ticket content")
+            logging.error(e.args)
+            return False
```
Is this `return False` designed as an error handling in the whole project? I would recommend have this consistent, as this does not seems to be the case, see:
```
grep "except requests.RequestException as e:" -A 3 ticketutil/*.py
```
I would recommend to reraise the error by `raise` instead of returning `False` and let the user decide what to do in case of error... @dmranck, what is your opinion on this?

dmranck replied:
@rpavelka, great point! In fact, that is the next thing on my To Do list for ticketutil - actually return valuable information from all methods - https://github.com/dmranck/ticketutil/issues/13

I'm planning to return a named tuple containing information like status of request, error message (if the request failed), ticket url (if a ticket was created), etc. And the user would simply access them with result.status, result.error_message, result.ticket_url, etc.

I'm hoping to get to this next month, and when I do, I'll be adding this functionality to ServiceNow as well.

pzubaty replied:
I see, it's only _verify* methods which return `False` in case of Exception. I had `raise` there, but I think there was comment from `@dmranck` he doesn't want re-raising Exceptions.

pzubaty commented:
I have addressed comments above and also question from Jakub Vulgan. There was a problem with Request POST query trying to use characters like `\n` or `"` without escaping them. That's what `json.dumps()` is for.

dmranck commented on docs/servicenow.md:
````diff
@@ -0,0 +1,173 @@
+t.remove_cc(['username@domain.com', 'user3@domain.com'])
+```
+
+### devops_one_url() <a name="devops_one"></a>
````
I still feel like we should remove the devops_one_url() stuff from ticketutil. Because it's a helper function for an internal instance of service now. Do you agree with this?

dmranck commented on ticketutil/servicenow.py:
```diff
@@ -0,0 +1,459 @@
+
+__author__ = 'dranck, rnester, kshirsal, pzubaty'
+
+STATE = {'new':'0',
```
Is this mapping just based off our internal SN instance, or is this consistent across all instances? Reason I'm asking is because for Redmine, I had to query the API to see what status ID corresponded to a status name. If SN allows for custom states, wouldn't the state's ID have to be retrieved somehow?

See https://github.com/dmranck/ticketutil/blob/master/ticketutil/redmine.py#L368 for Redmine _get_status_id() helper function.

pzubaty replied:
Good point, but this would need some investigation first.

pzubaty replied:
well, in the end I have replaced `STATE` with `self.available_states` which is being set up in `_verify_project()` This will be in new commit which is going to be pushed in 30 minutes.

dmranck commented on ticketutil/servicenow.py:
```diff
@@ -0,0 +1,459 @@
+        return fields
+
+
+    def devops_one_url(self):
```
I'm still not sure if this method should be in ticketutil. Because it's a Red Hat internal thing...

dmranck commented on ticketutil/servicenow.py:
```diff
@@ -0,0 +1,459 @@
+                server=self.url, sys_id=self.sys_id, table=self.project)
+
+
+def test():
```
It looks like test() and test_comment() are functional tests that you run to test out your code? I think these should be removed for now, and placed in a separate file that we don't commit to the ticketutil repo. I have files like this for all the other ticketutil tools, but I don't push them up to ticketutil because they contain Red Hat internal URLs and wouldn't apply to anyone else using the code.

Or we can keep them, but place them in a separate tests/ directory, similar to this: https://github.com/jstavel/ticketutil/blob/30498c0e2b683872b4bb75422f48c5dde4e77359/tests/test_bugzilla_attachments.py

I would like to roll out a functional test suite across all ticketutil tools at some point, but haven't started on this yet.

pzubaty replied:
You are right about these tests. I have forgotten to remove them and usually they are stored in my misc gitlab repo. Also I have moved devops_one_url function there.

dmranck commented on tests/test_servicenow.py:
```diff
@@ -0,0 +1,290 @@
+        mock_session.return_value = FakeSession()
+        ticket = servicenow.ServiceNowTicket(TEST_URL, TABLE,
+                                             ticket_id=TICKET_ID)
+        ticket.remove_cc(['dranck@redhat.com', 'dranck.com'])
```
Why include 'dranck.com' here? What about just 'dranck@redhat.com', or a second email adress, like 'mail@redhat.com'

dmranck commented on tests/test_servicenow.py:
```diff
@@ -0,0 +1,290 @@
+        mock_session.return_value = FakeSession(status_code=404)
+        ticket = servicenow.ServiceNowTicket(TEST_URL, TABLE,
+                                             ticket_id=TICKET_ID)
+        result = ticket.remove_cc(['dranck@redhat.com', 'dranck.com'])
```
Why include 'dranck.com' here? What about just 'dranck@redhat.com', or a second email adress, like 'mail@redhat.com'

pzubaty commented:
I have modified _verify_project so it queries sys_choice table and stores available ticket states (thanks to jhamann@redhat.com for the original query).

pzubaty commented:
Now I see that. I thought that query is returning what I need, but that's only order of states in the WebUI, but not their values.

pzubaty commented on docs/servicenow.md:
````diff
@@ -162,12 +162,3 @@ one email address.
 t.remove_cc(['username@domain.com', 'user3@domain.com'])
 ```
-
-### devops_one_url() <a name="devops_one"></a>
````
Done and done, I have renamed and left STATE in the unit tests, but it's using new mapping.
